### PR TITLE
fruit shop was implemented

### PR DESCRIPTION
**Pull request by not4und**

### CODE REVIEW

File: src/main/java/Main.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import service.strategy.impl.ReturnHandler;
+import service.strategy.impl.SupplyHandler;
+
+public class Main {
+    private static final String INPUT_CSV_PATH = "src/main/resources/input.csv";
+    private static final String REPORT_CSV_PATH = "src/main/resources/report.csv";
+    private static final FruitDao FRUIT_DAO = new FruitDaoImpl();
+    private static final ReaderService READER = new ReaderServiceImpl();
+    private static final ParserService PARSER = new ParserServiceImpl();
+    private static final WriterService WRITER = new WriteServiceImpl();
```
Comment (Rommelua) [inline src/main/java/Main.java:25]: These 4 fields are not constants, better make them local and create them in the main(). 
FYI [constant naming](https://mate-academy.github.io/style-guides/java/java.html#s5.2.4-constant-names)

File: src/main/java/Main.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    public static void main(String[] args) {
+        Map<Transaction.Operation, TransactionHandler> transactionMap = new HashMap<>();
+        transactionMap.put(Transaction.Operation.BALANCE, new BalanceHandler(FRUIT_DAO));
+        transactionMap.put(Transaction.Operation.PURCHASE, new PurchaseHandler(FRUIT_DAO));
+        transactionMap.put(Transaction.Operation.RETURN, new ReturnHandler(FRUIT_DAO));
+        transactionMap.put(Transaction.Operation.SUPPLY, new SupplyHandler(FRUIT_DAO));
```
Comment (Rommelua) [inline src/main/java/Main.java:32]: We should not create this map here in main()
This logic should be in a separate method 
or (better) create `public class TransactionExecutor` with Map inside and `public void execute(Transaction transaction)` method. In this `execute(Transaction transaction)` move this logic `transactionMap.get(transaction.getOperation()).apply(transaction))`

File: src/main/java/model/Transaction.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        Operation(String label) {
+            this.label = label;
+        }
+
+        public String getLabel() {
+            return label;
+        }
+
+        public static Operation valueOfLabel(String label) {
```
Comment (Rommelua) [inline src/main/java/model/Transaction.java:44]:
```suggestion
        public static Operation getByLabel(String label) {
```

File: src/main/java/model/Transaction.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            return Arrays.stream(values())
+                    .filter(o -> o.label.equals(label))
+                    .findFirst()
```
Comment (Rommelua) [inline src/main/java/model/Transaction.java:47]: Better use findAny() always when order does not matter.

File: src/main/java/service/ParserService.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package service;
+
+import java.util.List;
+import model.Transaction;
+
+public interface ParserService {
```
Comment (Rommelua) [inline src/main/java/service/ParserService.java:6]:
```suggestion
public interface TransactionParser {
```

File: src/main/java/service/ReaderService.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package service;
+
+import java.util.List;
+
+public interface ReaderService {
```
Comment (Rommelua) [inline src/main/java/service/ReaderService.java:5]: There is `Files.readAllLines()` in java core and I do not think we should reinvent the wheel creating our custom interface.

File: src/main/java/service/WriterService.java [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package service;
+
+public interface WriterService {
```
Comment (Rommelua) [inline src/main/java/service/WriterService.java:3]: The same as for reader - we can use Files class

File: src/main/java/service/strategy/TransactionHandler.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package service.strategy;
+
+import model.Transaction;
+
+public interface TransactionHandler {
+    void apply(Transaction transaction);
```
Comment (Rommelua) [inline src/main/java/service/strategy/TransactionHandler.java:6]: better call this method _execute_ or _run_

File: src/main/java/dao/FruitDao.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package dao;
+
+import java.util.Map;
+import model.Fruit;
+
+public interface FruitDao {
+    void update(Fruit fruit, Integer quantity);
+
+    Integer getQuantityOf(Fruit fruit);
```
Comment (Rommelua) [inline src/main/java/dao/FruitDao.java:9]:
```suggestion
    Integer getQuantity(Fruit fruit);
```

File: src/main/java/service/strategy/impl/PurchaseHandler.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        this.fruitDao = fruitDao;
+    }
+
+    @Override
+    public void apply(Transaction transaction) {
+        Fruit fruit = transaction.getFruit();
+        Integer fruitQuantity = fruitDao.getQuantityOf(fruit);
+        Integer transactionFruitQuantity = transaction.getQuantity();
+        if (fruitQuantity < transactionFruitQuantity) {
+            throw new RuntimeException("Not enough " + fruit.getName() + "s in stock.");
```
Comment (Rommelua) [inline src/main/java/service/strategy/impl/PurchaseHandler.java:21]: add actual and required quantity to message

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/service/impl/TransactionParserImpl.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    @Override
+    public List<Transaction> parse(List<String> inputList) {
+        inputList.remove(INFO_INPUT_LINE_INDEX);
+        List<Transaction> transactions = new ArrayList<>();
+        for (String input : inputList) {
+            String[] inputSplited = input.split(",");
+            transactions.add(new Transaction(
+                    Transaction.Operation.getByLabel(inputSplited[OPERATION_INDEX]),
+                    new Fruit(inputSplited[FRUIT_INDEX]),
+                    Integer.parseInt(inputSplited[QUANTITY_INDEX])
```
Comment (MaliukDaria) [inline src/main/java/service/impl/TransactionParserImpl.java:24]: as advise: move logic of creating transection into separate private method to increase code readability

File: src/main/java/service/impl/TransactionExecutorImpl.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+                break;
+            case PURCHASE:
+                new PurchaseHandler(fruitDao).execute(transaction);
+                break;
+            case SUPPLY:
+            case RETURN:
+                new AddHandler(fruitDao).execute(transaction);
+                break;
+            default:
+        }
```
Comment (MaliukDaria) [inline src/main/java/service/impl/TransactionExecutorImpl.java:27]: we should use "Strategy" pattern here with use of Map<Operation, OperationHandler>

File: src/main/java/Main.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+public class Main {
+    private static final String INPUT_CSV_PATH = "src/main/resources/input.csv";
+    private static final String REPORT_CSV_PATH = "src/main/resources/report.csv";
+
+    public static void main(String[] args) {
+        FruitDao fruitDao = new FruitDaoImpl();
+        TransactionParser transactionParser = new TransactionParserImpl();
+        TransactionExecutor transactionExecutor = new TransactionExecutorImpl();
+        try {
+            transactionParser.parse(Files.readAllLines(Path.of(INPUT_CSV_PATH)))
```
Comment (MaliukDaria) [inline src/main/java/Main.java:21]: We should move logic or reading/writing files in separate services

**Review by MaliukDaria — CHANGES_REQUESTED** [review]

File: src/main/java/Main.java (hunk 13 above)
Comment (not4und) [reply src/main/java/Main.java:21]: Another mentor says like "There is Files.readAllLines() in java core and I do not think we should reinvent the wheel creating our custom interface.". So i deleted Writer and Reader services.

**Review by aaananas — APPROVED** [review]

**Pull request closed without merging**